// @character-tech/client-common/src/utils/characterUtils.ts
import { Character, RecentChatShunt } from '../types/app-api';

const HiddenCharacterIds = [
  'T2j7WCwwEDJZod94YGY--FZjx0Ev-CfGTJPFfPt6jGY', // Creative Mode
];

export const filterHiddenCharacterChats = (chats: RecentChatShunt[]) => {
  if (HiddenCharacterIds.length === 0) {
    return chats;
  }
  return chats.filter(
    chat => HiddenCharacterIds.indexOf(chat.character_id) === -1,
  );
};

export const filterHiddenCharacters = (characters?: Character[]) => {
  if (HiddenCharacterIds.length === 0) {
    return characters;
  }
  return characters?.filter(
    character => HiddenCharacterIds.indexOf(character.external_id) === -1,
  );
};

// TODO: mobile needs to pull from these constants to get the most updated values
export const CharacterConstants = {
  MAX_BASE_PROMPT_LEN: 100,
  MIN_NAME_LEN: 3,
  MAX_NAME_LEN: 20,
  MIN_TITLE_LEN: 3,
  MAX_TITLE_LEN: 50,
  MIN_GREETING_LEN: 3,
  MAX_GREETING_LEN: 2048,
  MAX_DESCRIPTION_LEN: 500,
  MAX_DEFINITION_LEN: 32000,
  MAX_INTERACTIONS_BEFORE_EDITING_DISABLED: 1000000000,
};
