// @character-tech/client-common/src/utils/constants.ts
// Moderation constants.
export const IN_EU_KEY = 'in_eu';
export const IN_EEA_KEY = 'in_eea';
export const IN_UK_KEY = 'in_uk';
export const IP_API_URL = 'https://ipapi.co';
// ISO 3166-1 alpha-2 country code for the UK.
export const UK_COUNTRY_CODE = 'GB';
// ISO 3166-1 alpha-2 country codes that are in the EEA.
export const EEA_COUNTRY_CODES = [
  'AT', // Austria
  'BE', // Belgium
  'BG', // Bulgaria
  'HR', // Croatia
  'CY', // Cyprus
  'CZ', // Czech Republic
  'DK', // Denmark
  'EE', // Estonia
  'FI', // Finland
  'FR', // France
  'DE', // Germany
  'GR', // Greece
  'HU', // Hungary
  'IS', // Iceland
  'IE', // Ireland
  'IT', // Italy
  'LV', // Latvia
  'LI', // Liechtenstein
  'LT', // Lithuania
  'LU', // Luxembourg
  'MT', // Malta
  'NL', // Netherlands
  'NO', // Norway
  'PL', // Poland
  'PT', // Portugal
  'RO', // Romania
  'SK', // Slovakia
  'SI', // Slovenia
  'ES', // Spain
  'SE', // Sweden
];

export const DISCORD_HREF = 'https://discord.com/invite/characterai';
export const APP_STORE_HREF =
  'https://apps.apple.com/us/app/character-ai-ai-powered-chat/id1671705818';
export const GOOGLE_PLAY_STORE_HREF =
  'https://play.google.com/store/apps/details?id=ai.character.app';

export const validNameRegex =
  /^[\uAC00-\uD7AF\u4E00-\u9FFF\u3400-\u4DBF\u3040-\u30FF\u0400-\u04FFA-Za-z0-9áéíóúÁÉÍÓÚñÑâêîôûÂÊÎÔÛàèìòùÀÈÌÒÙçÇ_-][ \uAC00-\uD7AF\u4E00-\u9FFF\u3400-\u4DBF\u3040-\u30FF\u0400-\u04FFA-Za-z0-9áéíóúÁÉÍÓÚñÑâêîôûÂÊÎÔÛàèìòùÀÈÌÒÙçÇ_-]*$/i;
