// @character-tech/client-common/src/lib/axiosConstants.tsx
export const StagingNeoHost = 'neo-stage.characterai.dev';
export const DefaultNeoHost = 'neo.character.ai';

let host = DefaultNeoHost;
let modelServerURL: string | undefined = undefined; // model server override url
let websocketTtlMs = 15 * 60 * 1000; // 15 minutes
let websocketHeartbeatMs = 0; // disabled by default
let originId = '<not set>';
let debugMode = false;
let centrifugeTimeoutMs = 60000;
let eventSource: EventSource | undefined = undefined;

const protocolSecurity = () => (host.includes('localhost') ? '' : 's');

export const setNeoHost = (newHost: string) => (host = newHost);

export const setModelServerURL = (url: string) => (modelServerURL = url);

export const getHost = () => host;

export const getApiURL = () => `http${protocolSecurity()}://${host}/`;

export const getWSAddress = () => `ws${protocolSecurity()}://${host}/ws/`;

export const getCentrifugeWsAddress = () =>
  `ws${protocolSecurity()}://${host}/connection/websocket`;

export const getCentrifugeSseAddress = () =>
  `http${protocolSecurity()}://${host}/connection/sse`;

export const getCentrifugeEmulationAddress = () =>
  `http${protocolSecurity()}://${host}/connection/emulation`;

export const getModelServerURL = () => modelServerURL;

export const getWebsocketTtlMs = () => websocketTtlMs;

export const setWebsocketTtlMs = (newTtlMs: number) =>
  (websocketTtlMs = newTtlMs);

export const getOriginId = () => originId;

export const setOriginId = (newOriginId: string) => (originId = newOriginId);

export const getDebugMode = () => debugMode;

export const setDebugMode = (newDebugMode: boolean) =>
  (debugMode = newDebugMode);

export const getWebsocketHeartbeatMs = () => websocketHeartbeatMs;

export const setWebsocketHeartbeatMs = (newHeartbeatMs: number) =>
  (websocketHeartbeatMs = newHeartbeatMs);

export const getCentrifugeTimeoutMs = () => centrifugeTimeoutMs;

export const setCentrifugeTimeoutMs = (newCentrifugeTimeout: number) =>
  (centrifugeTimeoutMs = newCentrifugeTimeout);

export const setEventSource = (es: any) => (eventSource = es);

export const getEventSource = () => eventSource;

export const getMultimodalServerURL = () =>
  `https://${getHost()}/multimodal/api`;
